Add Reusable Block deletion

noisysocks opened the pull request:
## ✨ What this is

Special 2-for-1 offer: closes #3792 and closes #4041.

~~Note that I've set `add/reusable-block-deletion-to-api` as the base branch so that the diff is bearable. The intention is to merge this into `master` after https://github.com/WordPress/gutenberg/pull/4137 lands.~~

### 🗑 Adds the ability to delete a Reusable Block

This PR uses the new API introduced in https://github.com/WordPress/gutenberg/pull/4137 to add support for deleting Reusable Blocks.

With the new API, this is pretty straightforward:

1. There's a new `DELETE_REUSABLE_BLOCK` effect which calls the `destroy()` method that the WP API Backbone library gives us for free.
2. This effect optimistically calls `REMOVE_REUSABLE_BLOCK` which removes the Reusable Block and any blocks that reference it from local editor state. If the delete fails, `REMOVE_REUSABLE_BLOCK` is reverted.

###  💆‍♀️ Improves the Reusable Block flow

I implemented the changes suggested by @karmatosed in https://github.com/WordPress/gutenberg/issues/4041 by moving 'Detach Reusable Block' into a seperate settings menu section alongside our new 'Delete Reusable Block' button.

Here's how the settings menu now looks for a static block:

<img width="718" alt="screen shot 2017-12-21 at 12 29 19" src="https://user-images.githubusercontent.com/612155/34236295-f96982f0-e64a-11e7-9730-3d928d0e655c.png">

And here's how it now looks for a reusable block:

<img width="726" alt="screen shot 2017-12-21 at 12 29 46" src="https://user-images.githubusercontent.com/612155/34236297-fb00d654-e64a-11e7-9f5d-6a2468e4e6c2.png">

## 📋 How to test

1. Create a block
1. Convert it to a reusable block
1. Rename the reusable block
1. Edit the reusable block
1. Use the Inserter to insert a copy of the reusable block
1. Remove a block
1. Delete the reusable block

noisysocks committed:
Add DELETE_REUSABLE_BLOCK effect

Adds an effect (DELETE_REUSABLE_BLOCK) which triggers deletes a reusable
block via the API. The action that triggers this effect can be created
with the deleteReusableBlock() action creator.

The reusable block is optimistically removed from local redux state by
the reducer handling the REMOVE_REUSABLE_BLOCK action.

noisysocks committed:
Add UI for deleting a reusable block

Moves 'Detach from Reusable Block' and adds 'Delete Reusable Block'
buttons to a <ReusableBlockSettings> component that lives inside
<BlockSettingsMenu>.

noisysocks committed:
Rename 'Delete' button in block settings menu to 'Remove'

This helps reduce the ambiguity between removing a block and deleting a
Reusable Block.

noisysocks committed:
When deleting a Reusable Block, remove associated blocks from the post

Makes the REMOVE_REUSABLE_BLOCK action also remove any blocks that are
using the removed Reusable Block from the editor state.

noisysocks commented:
cc. @youknowriad @aduth @pento since they know this code
cc. @karmatosed @jasmussen for UX feedback

noisysocks committed:
Make success notice say 'Block created' when the Reusable Block is new

aduth commented on editor/components/block-settings-menu/reusable-block-settings.js:
```diff
@@ -0,0 +1,81 @@
+		onDelete( id ) {
+			// TODO: Make this a <Confirm /> component or similar
+			// eslint-disable-next-line no-alert
+			if ( window.confirm( __( 'Are you sure you want to permanently delete this Reusable Block?' ) ) ) {
```
You might be interested in a related discussion on "Scariness" of such prompts:

https://github.com/Automattic/wp-calypso/issues/10982

In other words, should we be proactive about messaging here to clarify that deleting a block can impact other posts?

noisysocks replied:
Good idea! Made the copy scarier in 6c0e18596d2a30558d4bb1a342d1d02259e9d44f.

aduth commented on editor/store/reducer.js:
```diff
@@ -759,6 +765,39 @@ export const reusableBlocks = combineReducers( {
+			case 'FETCH_REUSABLE_BLOCKS_FAILURE': {
+				const { id } = action;
+				if ( ! id ) {
+					return state;
```
In what situations would you expect this code path to be hit? Even if it was falsey, would there be harm in allowing `omit` to proceed? Aside from returning a new reference: But there's also an issue that if `id` is truthy but doesn't exist in the current set, we still return a new reference.

```js
omit( state, 'some-id-that-doesnt-exist' ) !== state
```

noisysocks replied:
You're right, this check isn't necessary. I've removed it in 92b6a18fc1f4b936358c4e0ddb8a9fc0957354aa.

aduth commented on editor/store/test/effects.js:
```diff
@@ -471,6 +474,8 @@ describe( 'effects', () => {
 
 	describe( 'reusable block effects', () => {
 		beforeAll( () => {
+			lodash.uniqueId = jest.spyOn( lodash, 'uniqueId' );
```
I wonder if we could achieve this more concisely with a combination of `jest.mock` and `require.requireActual`.

But then, looking at how you're asserting on `optimist.id`, do we really need this at all? Since you're using ` expect.any( Object ),` and not asserting against the actual value produced by `uniqueId`.

noisysocks replied:
You're right, mocking this isn't giving us much benefit for the effort involved. I've removed the mock in 78a9bb7657be82cc037d68ca121b27b6c68f75f7.

aduth reviewed (changes requested):
An edge case, but one that could be potentially destructive: We allow the user to delete a reusable block while it's still in its "temporary" unsaved state, meaning that it will use the [value produced by `uniqueId`](https://github.com/WordPress/gutenberg/blob/e2fb12740e6231b6d0f28a4fe95bd64cbe18d175/blocks/api/factory.js#L225). Since a block could exist with an ID matching the temporary ID, the user could inadvertently delete the wrong reusable block.

Luckily I didn't have a block with ID 2:

![image](https://user-images.githubusercontent.com/1779930/34501054-37748a7e-efdb-11e7-802c-02dfa5307d63.png)

I'd suggest:

- Prevent deleting reusable block if it hasn't been saved yet.
- Maybe pass the argument to [`uniqueId`](https://lodash.com/docs/4.17.4#uniqueId), which acts as a prefix. Arguably this shouldn't be necessary, but eliminates any chance that the temproary ID overlaps with a _real_ block ID.

chriskmnds commented on editor/store/effects.js:
```diff
@@ -362,15 +366,57 @@ export default {
+
+		const allBlocks = getBlocks( getState() );
+		const associatedBlocks = allBlocks.filter( block => isReusableBlock( block ) && block.attributes.ref === id );
+		const associatedBlockUids = associatedBlocks.map( block => block.uid );
```
Might a reducer here (i.e. `allBlocks.reduce...`) help avoid passing through the blocks list twice? Or is that a heavier operation? Just a tiny thought on optimisation I guess :-)

noisysocks replied:
You're absolutely right that using `reduce` would likely be faster, but I prefer to optimise for readability over performance unless it's an operation that's performance sensitive. Deleting a reusable block is a fairly uncommon operation, so I'm not very worried.

noisysocks committed:
Prevent temporary reusable blocks from being deleted

Don't allow DELETE_REUSABLE_BLOCK to delete a reusable block that is
marked temporary.

noisysocks committed:
Make Reusable Block delete confirmation scarier

noisysocks committed:
Simplify `reusableBlocks.isFetching` reducer

There's no need to check for the existence of `id` before `omit`ing it.
If `id` is falsey then `omit` will do nothing.

noisysocks committed:
Don't mock lodash.uniqueId

We can avoid mocking `uniqueId` by being less pedantic about what our
tests expect.

noisysocks commented:
Thanks for reviewing, @aduth! I've addressed your comments and made it so that reusable blocks with temporary IDs cannot be deleted in 1f9e5c24d899cb2880e33647c481c5cea859f6e6.

aduth reviewed (approved):
Nice work 👍 

Question: Should the deleted block be removed from `state.preferences.blockUsage` as well? I've noticed there's some issues with Reusable Blocks in recent and frequently used blocks, so wasn't sure if you planned to address this separately. I'd guess that the reusable block of `state.preferences.blockUsage` is not storing the `ref` anyways, so not tracking details of the block being removed.

noisysocks commented:
> Question: Should the deleted block be removed from state.preferences.blockUsage as well? 

That's a good point. I'll do this as part of https://github.com/WordPress/gutenberg/pull/4224 since, as you say, we aren't currently tracking reusable blocks in the frequent/recent list of blocks.
